controllers/deployment_controller.go

```diff
@@ -106,23 +114,108 @@ func (r *DeploymentReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
 }
 
 func (r *DeploymentReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		WithOptions(controller.Options{
+	c, err := controller.New("csi-controller", mgr,
+		controller.Options{
+			Reconciler: r,
 			// only one instance of CSIDeployment is allowed to be installed
 			// concurrent reconciliation isn't supported
 			MaxConcurrentReconciles: 1,
-		}).
-		Watches(&source.Kind{Type: &csibaremetalv1.Deployment{}}, &handler.EnqueueRequestForObject{}).
-		Watches(&source.Kind{Type: &appsv1.Deployment{}}, &handler.EnqueueRequestForOwner{
-			IsController: true,
-			OwnerType:    &csibaremetalv1.Deployment{},
-		}).
-		Watches(&source.Kind{Type: &appsv1.DaemonSet{}}, &handler.EnqueueRequestForOwner{
-			IsController: true,
-			OwnerType:    &csibaremetalv1.Deployment{},
-		}).
-		For(&csibaremetalv1.Deployment{}).
-		Complete(r)
+		})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &csibaremetalv1.Deployment{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &appsv1.Deployment{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &csibaremetalv1.Deployment{},
+	})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &appsv1.DaemonSet{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &csibaremetalv1.Deployment{},
+	})
+	if err != nil {
+		return err
+	}
+
+	// reconcile CSI Deployment if node was creates, node kernel-version or label were changed
+	err = c.Watch(&source.Kind{Type: &corev1.Node{}}, handler.EnqueueRequestsFromMapFunc(func(a client.Object) []reconcile.Request {
+		ctx := context.Background()
+		deployments := &csibaremetalv1.DeploymentList{}
+		err := r.Client.List(ctx, deployments)
+		if err != nil {
+			return []reconcile.Request{}
+		}
+
+		var requests []reconcile.Request
+		for _, dep := range deployments.Items {
+			requests = append(requests, reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name:      dep.Name,
+					Namespace: dep.Namespace,
+				}})
+		}
+
+		return requests
+	}),
+		predicate.Funcs{
+			CreateFunc: func(createEvent event.CreateEvent) bool {
+				return true
```

Comment:
Does this filter work by default?

Reply:
Yes, you're right. I expected only events, which we set here, would work. I deleted this filter and changed only UpdateFunc